pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -498,12 +491,20 @@ func getTPMapMatchingIncomingAffinityAntiAffinity(pod *v1.Pod, nodeInfoMap map[s
 				}
 			}
 		}
+
 		if len(nodeTopologyPairsAffinityPodsMaps.topologyPairToPods) > 0 || len(nodeTopologyPairsAntiAffinityPodsMaps.topologyPairToPods) > 0 {
 			appendResult(node.Name, nodeTopologyPairsAffinityPodsMaps, nodeTopologyPairsAntiAffinityPodsMaps)
 		}
 	}
 	workqueue.ParallelizeUntil(ctx, 16, len(allNodeNames), processNode)
-	return topologyPairsAffinityPodsMaps, topologyPairsAntiAffinityPodsMaps, firstError
+
+	select {
+	case err := <-errCh:
+		return nil, nil, err
+
+	default:
+		return topologyPairsAffinityPodsMaps, topologyPairsAntiAffinityPodsMaps, nil
+	}
```

Comment:
To potentially address @bsalamat concern about readability, and help establish a clear pattern for signaling errors, we could make capturing the error a function as well, and by that we have a clearer and established error send/receive pattern.

1) Create a function for receiving the error similar to send in utils.go:
```
func NonBlockingReceiveError(errCh chan<- error) error {
  select {
	case err := <-errCh:
		return err
	default:
  }
  return nil
}
```

2) here we just call the function:
```
if err := schedulerutil.NonBlockingReceiveError(errCh); err != nil {
 return nil, nil, err
}
return topologyPairsAffinityPodsMaps, topologyPairsAntiAffinityPodsMaps, nil
```

This NonBlockingSendError/NonBlockingReceiveError pattern can now be used in all places where we need to capture a single error from parallel loops.

Reply:
What does an `ErrorRecorder` with the following APIs sounds, it encapsulates related functions and provides an approach to collect errors?

```go
func NewErrorRecorder() *ErrorRecorder
func (r *ErrorRecorder) Record(err error) bool
func (r *ErrorRecorder) FirstError() error
```

Reply:
Adding an explicit abstraction sounds good to me, this is even better.

I would not call prefix it with Channel since that exposes the implementation detail of the abstraction. I would suggest two modifications: 

1) give the option to pass a callback to NewErrorRecorder so that it gets called by send error if set. 
2) change error to interface because in some places we have other ways to report errors, like the framework Status.

```
func NewErrorRecorder(bufferSize int, cancel context.CancelFunc) *ErrorRecorder

func (c *ErrorChannel) SendError(err interface{}) bool

func (c *ErrorChannel) ReceiveError() interface{}
```

Reply:
> Adding an explicit abstraction sounds good to me, this is even better.
> 
> I would not call suffix it with Channel since that exposes the implementation detail of the abstraction. I would suggest two modifications:
> 
> 1. give the option to pass a callback to NewErrorRecorder so that it gets called by send error if set.
> 2. change error to interface because in some places we have other ways to report errors, like the framework Status.
> 
> ```
> func NewErrorRecorder(bufferSize int, cancel context.CancelFunc) *ErrorRecorder
> 
> func (c *ErrorChannel) SendError(err interface{}) bool
> 
> func (c *ErrorChannel) ReceiveError() interface{}
> ```

Change the type to interface{} would cause an extra explicit conversion to error which looks bad...

```go
recorder.RecieverError().(error)
```

Reply:
Hi @bsalamat @Huang-Wei, could you give us some suggestions on this? Do we need to `ErrorRecorder` to encapsulate methods and if it sounds good, how should we define the API it provided?

Reply:
I found similar logic here in these places which do not need to call `CancelFunc`, and I still think it should not be part of the API.

+ https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/priorities/interpod_affinity.go#L194-L218 
+ https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L760-L782
+ https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1563-L1596

And we probably go too far on this, use select and channel to capture the first error is quite common in the codebase, I use following cmd and find many similar structure.

```go
$ ag "case.*errCh.*"
...
```

Reply:
> Can't they pass nil if they don't want to pass a callback?

Since it is not a typical pattern, it would be better to keep the API simple instead of adding support for exceptional cases. And call cancel function separately would work too.

Reply:
Sorry, I am not sure what you mean by "since it is not a typical pattern". 

If you mean it is not common that we pass a callback in general, then I guess we discussed that in the other PR and you seem to have agreed that it is a common pattern to pass callbacks, and I pointed that cancel is callback.

If you mean it is not common in our code base that we need to call cancel, then that is not true, you have four cases in this PR, I have one in my PR. Also, from the four cases you pointed out, one should be changed to use cancel, the other isn't in scope since it is not a parallel loop:

> > * https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/priorities/interpod_affinity.go#L194-L218

This should be changed to use cancel.

> > * https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L760-L782


This is not a case that we need to cover, this isn't a loop executed in parallel.


If you are concerned about passing nil, we could have a new function that allows setting the callback:

```
func NewErrorRecorderWithCancel(bufferSize int, cancel context.CancelFunc) *ErrorRecorder
```